### PR TITLE
[DG] Update find command and add Cancelled Dates

### PR DESCRIPTION
**Pull request by lingshanng**

### CODE REVIEW

File: docs/DeveloperGuide.md [hunk 1]
```diff
@@ -520,27 +546,27 @@ Figure I.4.1 shows a sequence diagram of how viewing tags works.<br>
...
-**Current Implementation**
+Finding students is facilitated by the `PersonMatchesKeywordsPredicate`.
+* `PersonMatchesKeywordsPredicate` stores a list of keywords for each field that is being searched for.
+* It also stores a `FindCondition` which takes on one of the values `ALL`, `ANY`, or `NONE`, and defaults to `ALL`. It determines whether all, any or none of the fields specified are required to match with a given person, for the person to be returned.
+
+Given below is an example usage scenario and how the Find Command is executed:
+* **Step 1:** The user enters the `find` command with name keyword `John` and tag keyword `unpaid`.
+* **Step 2:** The `FindCommandParser` parses this command and creates a `PersonMatchesKeywordsPredicate` that stores the respective keywords for name and tag.
+* **Step 3:** This predicate is passed into the method `Model#updateFilteredPersonList`.
+* **Step 4:** When the `PersonMatchesKeywordsPredicate#test()` method is called, a predicate is created for each searched field, name and tag, which tests if the given person's fields contain all the specified keywords.
```
Comment (Xiaoyunnn) [inline docs/DeveloperGuide.md:559]: Should it be `PersonMatchesKeywordsPredicate#test(Person)` ?

File: docs/diagrams/FindUpdateFilteredListLogicSequenceDiagram.puml [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+LogicManager -> FindCommand : execute()
+activate FindCommand
+
+FindCommand -> Model : updateFilteredPersonList(p)
+activate Model
+
+Model --> FindCommand
+deactivate Model
+
+Model -> PersonMatchesKeywordsPredicate : test
```
Comment (Xiaoyunnn) [inline docs/diagrams/FindUpdateFilteredListLogicSequenceDiagram.puml:27]: `test(p)`?

File: docs/DeveloperGuide.md [hunk 3]
```diff
@@ -183,8 +183,8 @@ This section describes some noteworthy details on how certain features are imple
 Weekly recurring or one-off (makeup) lessons are classified as `Lesson` objects. These lessons can be added to any particular
 student in TAB. Added lessons can also be edited and deleted.<br>
 
-A `Lesson` is represented in the application as shown in the figure below. It contains a start `Date`, an end `Date`, a `TimeRange` for the
-`Lesson`, a `Subject`, a `LessonRates` and `Homework` fields. There are 2 types of `Lesson` – `RecurringLesson` and `MakeUpLesson`. `RecurringLesson`
+A `Lesson` is represented in the application as shown in the figure below. It contains a start `Date`, an end `Date`, a `CancelledDates` set,  a `TimeRange` for the
+`Lesson`, , a `Subject`, a `LessonRates` and `Homework` fields. There are 2 types of `Lesson` – `RecurringLesson` and `MakeUpLesson`. `RecurringLesson`
```
Comment (eeliana) [inline docs/DeveloperGuide.md:187]:
```suggestion
`Lesson`, a `Subject`, a `LessonRates` and `Homework` fields. There are 2 types of `Lesson` – `RecurringLesson` and `MakeUpLesson`. `RecurringLesson`
```

File: docs/DeveloperGuide.md [hunk 4]
```diff
@@ -256,6 +256,12 @@ The `executeUndoableCommand()` method of the `LessonEditCommand` uses this `edit
 The new lesson is stored in TAB in place of the old lesson. The student's list of lessons will be updated to reflect
 the changes made to the specified lesson.<br>
 
+#### Cancelling and Uncancelling Lessons
+
+Cancelling and uncancelling of lessons is done using the `LessonEditCommand`, which allows the user to specify multiple cancel and uncancel dates in one command. 
+
+Cancelled dates of a lesson are stored as a field, `CancelledDates` in the lesson object. When a user specifies a valid lesson date to cancel, the date is added into the `CancelledDates` set. When a user specifies a valid lesson date to uncancel, the date is removed from `CancelledDates`.
```
Comment (eeliana) [inline docs/DeveloperGuide.md:263]:
```suggestion
Cancelled dates of a lesson are stored as a field, `CancelledDates`, in the lesson object. When a user specifies a valid lesson date to cancel, the date is added into the `CancelledDates` set. When a user specifies a valid lesson date to uncancel, the date is removed from `CancelledDates`.
```

File: docs/DeveloperGuide.md [hunk 5]
```diff
@@ -360,6 +366,26 @@ We chose alternative 2 and integrated CalendarFX into our app as the possibility
...
+
+![Recurring Calendar Entries](images/RecurringCalendarEntries.png)
+*Figure I.1.2: Recurring lesson displayed on the yearly calendar*
+
+To represent this recurring lesson, 3 calendar entries are created:
+1. Recurring entry 1 with start date of `1 Sep 2021`, and end date of `15 Sep 2021`
+2. Recurring entry 2 with start date of `29 Sep 2021` and end date of `6 Oct 2021`
+3. Recurring entry 3 with start date of `20 Oct 2021` and no end date.
+
+In this way, the lesson will not be  displayed on the cancelled dates.
```
Comment (eeliana) [inline docs/DeveloperGuide.md:387]:
```suggestion
In this way, the lesson will not be displayed on the cancelled dates.
```

File: docs/DeveloperGuide.md [hunk 6]
```diff
@@ -262,6 +262,12 @@ The `executeUndoableCommand()` method of the `LessonEditCommand` uses this `edit
 The new lesson is stored in TAB in place of the old lesson. The student's list of lessons will be updated to reflect
 the changes made to the specified lesson.<br>
 
+#### Cancelling and Uncancelling Lessons
+
+Cancelling and uncancelling of lessons is done using the `LessonEditCommand`, which allows the user to specify multiple cancel and uncancel dates in one command.
```
Comment (eeliana) [inline docs/DeveloperGuide.md:267]:
```suggestion
Cancelling and uncancelling of lessons is done using the `LessonEditCommand`, which allows the user to specify multiple dates to be cancelled and uncancelled dates in one command. 
```

**Comment by lingshanng** [comment]
wait, I'm going to add the part for Efforts

File: docs/DeveloperGuide.md [hunk 7]
```diff
@@ -1415,6 +1434,8 @@ Implementation of these features took ~4k Loc.
 
 **Finding students**
 
+The finding students feature required much consideration in designing the find constraints to ensure it is flexible and usable for users. It was challenging to implement a data structure to allow searching by multiple fields concurrently, which may each accept multiple keywords, and may have different search behaviours. In addition to student fields, lesson fields were also used in the find feature, increasing its complexity. The find feature also allows users to optionally customise the matching condition, which required more effort to implement. The large number of parameters and variations also meant that it was more susceptible to bugs, and it required intensive and careful testing.
```
Comment (eeliana) [inline docs/DeveloperGuide.md:1437]:
```suggestion
The finding students feature required much considerations in designing the find constraints to ensure it is flexible and usable for users. It was challenging to implement a data structure to allow searching by multiple fields concurrently, which may each accept multiple keywords, and may have different search behaviours. In addition to student fields, lesson fields were also used in the find feature, increasing its complexity. The find feature also allows users to optionally customise the matching condition, which required more effort to implement. The large number of parameters and variations also meant that it was more susceptible to bugs, and it required intensive and careful testing. 
```

**Review by eeliana — APPROVED** [review]

**Pull request merged**